File: src/unity/Runtime/Core/ObserverManager.cs
using System;
using System.Collections.Generic;

namespace EE {
    public class ObserverManager<Observer> : IObserverManager<Observer> {
        private int _counter;
        private readonly Dictionary<int, Observer> _observers;

        public ObserverManager() {
            _counter = 0;
            _observers = new Dictionary<int, Observer>();
        }

        public int AddObserver(Observer observer) {
            var id = _counter++;
            _observers.Add(id, observer);
            return id;
        }

        public bool RemoveObserver(int id) {
            return _observers.Remove(id);
        }

        public void DispatchEvent(Action<Observer> dispatcher) {
            foreach (var entry in _observers) {
                dispatcher(entry.Value);
            }
        }
    }
}

File: src/ts/src/core/ObserverManager.ts
import { IObserverManager } from "./IObserverManager";

export class ObserverManager<Observer> implements IObserverManager<Observer> {
    private _counter = 0;
    private readonly _observers: { [index: number]: Observer } = {};

    public addObserver(observer: Observer): number {
        const id = this._counter++;
        this._observers[id] = observer;
        return id;
    }

    public removeObserver(id: number): boolean {
        if (this._observers[id]) {
            delete this._observers[id];
            return true;
        }
        return false;
    }

    public dispatchEvent(dispatcher: (observer: Observer) => void): void {
        for (const key in this._observers) {
            const observer = this._observers[key];
            dispatcher(observer);
        }
    }
}